### PR TITLE
[CDAP-20381] Add ability to configure java options for program run

### PR DESCRIPTION
**Pull request by rmstar**

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/AbstractRuntimeTwillPreparer.java [hunk 1]
```diff
@@ -89,6 +91,7 @@
 import org.apache.twill.internal.json.ArgumentsCodec;
 import org.apache.twill.internal.json.TwillRuntimeSpecificationAdapter;
 import org.apache.twill.internal.utils.Paths;
+import org.elasticsearch.common.Strings;
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/AbstractRuntimeTwillPreparer.java]: use guava Strings instead

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/DistributedProgramRunner.java [hunk 2]
```diff
@@ -351,21 +352,24 @@ public final ProgramController run(final Program program, ProgramOptions oldOpti
           twillPreparer.setSchedulerQueue(schedulerQueueName);
         }
 
-        // Set JVM options based on configuration
+        // Set JVM options based on configuration. Skip adding JVM options for tethering provisioner as they are
+        // added in AbstractRuntimeTwillPreparer's constructor
         String jvmOpts = cConf.get(Constants.AppFabric.PROGRAM_JVM_OPTS);
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/DistributedProgramRunner.java:368]: I would have thought things would work if we just look in the options for any additional jvm options and add them to the twillPreparer here. Not clear on why we need all these other changes and why we need to special case tethering.

Comment (rmstar) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/DistributedProgramRunner.java:368]: > we just look in the options for any additional jvm options and add them to the twillPreparer here

I will try this and update here.

> why we need to special case tethering.

Couple of reasons to special case tethering:
1. Avoid adding jvm options that are not relevant to running on k8s - -DCDAP_LOG_DIR=, -Dlogback.configurationFile==resources.jar/resources/logback.xml, -Dtwill.container.class.loader
2. Avoid adding `app.program.jvm.opts` twice to the twillPreparer. It's first added in the twillPreparer's [constructor](https://github.com/cdapio/cdap/blob/3435c5a5d3f2e975792336a8d64aef1a85092a5e/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/AbstractRuntimeTwillPreparer.java#L145). So avoid re-adding `app.program.jvm.opts` to the twillPreparer here.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/AbstractRuntimeTwillPreparer.java (hunk 1 above)
Comment (rmstar) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/AbstractRuntimeTwillPreparer.java]: Thanks, will fix

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/DistributedProgramRunner.java (hunk 2 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/DistributedProgramRunner.java:368]: Is it already getting added twice? If so it seems like it shouldn't be getting added in the constructor?

Comment (albertshau) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/DistributedProgramRunner.java:368]: for the first point about logback, is this an existing issue? If so it would be better to do that in a separate PR rather than mix up multiple fixes in this one.

Comment (rmstar) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/DistributedProgramRunner.java:368]: - removed jvm options addition in constructor
- reverted the logback change

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/AbstractRuntimeTwillPreparer.java [hunk 3]
```diff
@@ -773,7 +777,7 @@ private JvmOptions getJvmOptions() {
    * Returns the extra options for the container JVM.
    */
   private String addClassLoaderClassName(String extraOptions) {
-    if (classLoaderClassName == null) {
+    if (classLoaderClassName == null || TetheringProvisioner.TETHERING_NAME.equals(provisioner)) {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/AbstractRuntimeTwillPreparer.java]: why is this needed?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/RuntimeJobTwillPreparer.java [hunk 4]
```diff
@@ -133,49 +129,7 @@ private RuntimeJobInfo createRuntimeJobInfo(RuntimeSpecification runtimeSpec,
...
-    return new DefaultRuntimeJobInfo(getProgramRunId(), resultingFiles, jvmProperties);
+    return new DefaultRuntimeJobInfo(getProgramRunId(), resultingFiles, jvmOpts);
   }
 
-  /**
-   * Parses a jvm options string (e.g. {@code -XYZ -Dkey=value}) and extracts key/value properties
-   * provided by the {@code -D} options.
-   */
-  @VisibleForTesting
-  static Map<String, String> parseJvmProperties(String jvmOpts) {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/RuntimeJobTwillPreparer.java]: why do we need to move this code from here to the RuntimeJob?

the reason I ask is that this class is used on the instance side, before the job is submitted, whereas RuntimeJob is executed on the cluster. So if there is an invalid value it would be better to fail earlier.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/AbstractRuntimeTwillPreparer.java [hunk 5]
```diff
@@ -139,10 +143,10 @@ abstract class AbstractRuntimeTwillPreparer implements TwillPreparer {
     this.twillSpec = twillSpec;
     this.programRunId = programRunId;
     this.programOptions = programOptions;
+    this.provisioner = SystemArguments.getProfileProvisioner(programOptions.getArguments().asMap());
     this.locationCache = locationCache;
     this.locationFactory = locationFactory;
     this.controllerFactory = controllerFactory;
-    this.extraOptions = cConf.get(io.cdap.cdap.common.conf.Constants.AppFabric.PROGRAM_JVM_OPTS);
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/AbstractRuntimeTwillPreparer.java:145]: do you know why this was here to begin with? Does it mean we were double-setting these jvm options this whole time?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/AbstractRuntimeTwillPreparer.java (hunk 3 above)
Comment (rmstar) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/AbstractRuntimeTwillPreparer.java]: This is so that we don't do add `-Dtwill.container.class.loader` to jvm opts for tethered runs (I don't think adding it causes an issues; it's just unnecessary).

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/AbstractRuntimeTwillPreparer.java (hunk 5 above)
Comment (rmstar) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/AbstractRuntimeTwillPreparer.java:145]: Not sure why this was here to begin with. 
We started setting jvm options in DistributedProgramRunner as part of https://github.com/cdapio/cdap/pull/13226.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/RuntimeJobTwillPreparer.java (hunk 4 above)
Comment (rmstar) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/RuntimeJobTwillPreparer.java]: I moved parseJvmProperties() to DefaultRuntimeJobInfo (not DefaultRuntimeJob), which is called from RuntimeJobTwillPreparer -- see line 132 above.
Reason for doing this: DefaultRuntimeJobInfo now needs to keep track of both the raw jvm options (string) as well as parsed jvm properties (map of key/value pairs).

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/AbstractRuntimeTwillPreparer.java (hunk 3 above)
Comment (rmstar) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/AbstractRuntimeTwillPreparer.java]: Before this PR, TetheringRuntimeJobManager just copied over `app.program.jvm.opts` from cConf - https://github.com/cdapio/cdap/blob/develop/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/runtime/spi/runtimejob/TetheringRuntimeJobManager.java#L257

Now, we are setting `app.program.jvm.opts` in the TetheringLaunchMessage using the jvm opts added in the twill preparer.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/AbstractRuntimeTwillPreparer.java (hunk 5 above)
Comment (rmstar) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/AbstractRuntimeTwillPreparer.java:145]: > Does it mean we were double-setting these jvm options this whole time?

Yes, I think so

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/AbstractRuntimeTwillPreparer.java (hunk 3 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/AbstractRuntimeTwillPreparer.java]: ok I see, it's used by DataprocJobMain and RemoteLauncher, which isn't used in the tethering flow. I don't think we actually need to skip this, but if we do, it should be done by skipping the call to `.setClassLoader()` on the TwillPreparer when there is a tethered run, instead of skipping adding it in this class. Otherwise we're relying on an implementation detail to avoid setting it rather than using the actual API.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/AbstractRuntimeTwillPreparer.java (hunk 5 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/AbstractRuntimeTwillPreparer.java:145]: ok I see

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/runtimejob/DefaultRuntimeJobInfo.java [hunk 6]
```diff
@@ -46,7 +51,9 @@ public DefaultRuntimeJobInfo(ProgramRunId programRunId, Collection<? extends Loc
         .setProgram(programRunId.getProgram())
         .setRun(programRunId.getRun()).build();
     this.files = Collections.unmodifiableCollection(new ArrayList<>(files));
-    this.jvmProperties = Collections.unmodifiableMap(new LinkedHashMap<>(jvmProperties));
+    this.jvmOpts = jvmOpts;
+    this.jvmProperties = Collections.unmodifiableMap(parseJvmProperties(jvmOpts));
+    LOG.info("JVM properties {}", jvmProperties);
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/runtimejob/DefaultRuntimeJobInfo.java]: In general, logging in a constructor is usually a bad idea, tends to result in duplicate logging.

If we want to log, it should be close to where the jvm properties are actually being used

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/RuntimeJobTwillPreparer.java (hunk 4 above)
Comment (albertshau) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/RuntimeJobTwillPreparer.java]: ah misread the classname

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/runtime/spi/runtimejob/TetheringRuntimeJobManager.java [hunk 7]
```diff
@@ -248,14 +249,15 @@ private TetheringLaunchMessage createLaunchPayload(RuntimeJobInfo runtimeJobInfo
...
-          .findFirst().orElseThrow(() -> new IllegalStateException("Cannot find file" + fileName));
+          .findFirst().orElseThrow(() -> new IllegalStateException("Cannot find file " + fileName));
       builder.addLocalizeFiles(fileName, getLocalFileAsCompressedBytes(localFile));
     }
 
     for (String prefix : SELECT_CCONF_FIELDS) {
       String prefixRegex = "^" + prefix.replace(".", "\\.");
       builder.addCConfEntries(cConf.getValByRegex(prefixRegex));
     }
+    builder.addCConfEntry(Constants.AppFabric.PROGRAM_JVM_OPTS, runtimeJobInfo.getJvmOpts());
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/runtime/spi/runtimejob/TetheringRuntimeJobManager.java]: I'm still not clear why this is needed when we're already sending the program options over. Doesn't it still go through the DistributedProgramRunner eventually on the tethered side where the options get added to the TwillPreparer?

In other words, what breaks if we just do the change in the DistributedProgramRunner? Does the tethered side go through a different code path for some reason?

Comment (rmstar) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/runtime/spi/runtimejob/TetheringRuntimeJobManager.java]: Oh you're right, thanks for pointing that out! I was testing with this change (add jvm opts to twill preparer) applied only on the CDF side, so the updated jvm opts didn't take effect for tethered runs.
When this fix is applied on both sides, the cConf update is unnecessary - fixed now. Thank you!

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/runtimejob/DefaultRuntimeJobInfo.java (hunk 6 above)
Comment (rmstar) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/runtimejob/DefaultRuntimeJobInfo.java]: reverted this change

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/AbstractRuntimeTwillPreparer.java (hunk 3 above)
Comment (rmstar) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/AbstractRuntimeTwillPreparer.java]: reverted this change.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/AbstractRuntimeTwillPreparer.java [hunk 8]
```diff
@@ -18,6 +18,7 @@
 
 import com.google.common.base.Joiner;
 import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/AbstractRuntimeTwillPreparer.java]: don't need this anymore

**Review by albertshau — APPROVED** [review]
